### ISSUE TITLE
test_simd_int32x4 fails on current spidermonkey

### ISSUE THREAD

**Issue by kripken** [kripken/emscripten#4165]
diff between node.js and spidermonkey (inbound from a few days ago):

````
diff n m
21,22c21,22
< emscripten_int32x4_shiftLeftByScalar(v, 32): 0 0 0 0
< emscripten_int32x4_shiftLeftByScalar(v, 48): 0 0 0 0
---
> emscripten_int32x4_shiftLeftByScalar(v, 32): 1 0 1 4
> emscripten_int32x4_shiftLeftByScalar(v, 48): 65536 0 65536 262144
27c27
< emscripten_int32x4_shiftRightByScalar(v, 32): 0 0 0 0
---
> emscripten_int32x4_shiftRightByScalar(v, 32): 1 0 1 4
````

cc @juj @stoklund

**Comment by stoklund** [comment]
The SIMD.js semantics for oversized shifts recently changed, and SpiderMonkey implements the new semantics.

The shift amount is reduced modulo the bit-size of a lane, so shifting an int32 by 32 is now the same as shifting by 0.

**Comment by kripken** [comment]
Oh ok, thanks. Sounds like we need to update that test then.

Do you know if the polyfill has been updated so that it would produce the same output as SpiderMonkey? If so then we should update the polyfill as well.

**Comment by stoklund** [comment]
The polyfill was updated by https://github.com/tc39/ecmascript_simd/commit/d08bcc78088196ca974a8fee325e89d2afdab997

**Comment by kripken** [comment]
Thanks, updated the polyfill and the tests.

@juj, can you please update spidermonkey on the bots when you get a chance? I think those tests will break on that until it's updated. No rush though.

Let's leave the issue open for that.

**Comment by kripken** [comment]
@juj, that commit also broke `test_sse2_full`. I need your help there, not sure what the right fix is, since that uses existing APIs, not our own `emscripten_*` stuff.

**Comment by kripken** [comment]
post-gdc ping, @juj - this is broken on the bots, and I'm not familiar enough with native SIMD APIs to know what we should do.

**Comment by juj** [comment]
Working on this now..

**Comment by juj** [comment]
Fixed in above commit.

**Comment by kripken** [comment]
Thanks!

So there is no fast way to implement those APIs? Should we tell people not to use them?

**Comment by juj** [comment]
The shift instructions are only part of a larger set of emulated operations that are not fast (min/max are in this category, as are anything float64x2, int64x2, int128x1, saturating arithmetic, averaging, conversions, multiply-add, substract-add, mulhi, ...). I estimate overall 60%-70% of SSE instructions to be emulated.

At some point I was arguing a lot regarding individual SIMD instructions, but the feedback I got was very holistic about data on "does SIMD still give an overall performance win?" vs caring about individual instructions, so generally I've followed this stance to not document out individual emulated instructions, but message to developers to try SIMD out overall, and if it helps in the codebase they are building, then use it, but if not, then forget it and keep the eye for SIMD.js v2 later where we'd be looking towards having more instructions in the mix.
